File: android/app/src/main/java/com/gooseco/myliftsquad/data/api/OplApiService.kt
package com.gooseco.myliftsquad.data.api

import com.google.gson.JsonArray
import com.google.gson.JsonParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

data class SearchPage(
    val athletes: List<OplAthlete>,
    val nextMenStart: Int?,
    val nextWomenStart: Int?
) {
    val hasMore: Boolean get() = nextMenStart != null || nextWomenStart != null
}

class OplApiService {

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val baseUrl = "https://www.openpowerlifting.org/api"

    /**
     * Search both men and women paths in parallel, combine and deduplicate by slug.
     * Pass nextMenStart/nextWomenStart from a previous SearchPage to load more results.
     * A null start means that gender path is exhausted and will be skipped.
     */
    suspend fun searchAthletes(
        query: String,
        menStart: Int? = 0,
        womenStart: Int? = 0
    ): SearchPage = coroutineScope {
        val menDeferred = if (menStart != null) {
            async { searchPath("men", query, menStart) }
        } else {
            async { Pair(emptyList<OplAthlete>(), null) }
        }
        val womenDeferred = if (womenStart != null) {
            async { searchPath("women", query, womenStart) }
        } else {
            async { Pair(emptyList<OplAthlete>(), null) }
        }

        val (menAthletes, nextMenStart) = menDeferred.await()
        val (womenAthletes, nextWomenStart) = womenDeferred.await()

        val combined = LinkedHashMap<String, OplAthlete>()
        for (athlete in menAthletes + womenAthletes) {
            combined.putIfAbsent(athlete.slug, athlete)
        }
        val queryLower = query.trim().lowercase()
        val filtered = combined.values.filter { it.name.lowercase().contains(queryLower) }

        SearchPage(filtered, nextMenStart, nextWomenStart)
    }

    /**
     * Fetch full competition history for an athlete by slug.
     * Uses /api/liftercsv/{slug} which returns CSV with all entries sorted newest first.
     */
    suspend fun fetchCompetitionHistory(slug: String): List<CompetitionResult> =
        withContext(Dispatchers.IO) {
            try {
                val csv = get("$baseUrl/liftercsv/$slug") ?: return@withContext emptyList()
                parseCsv(csv)
            } catch (e: Exception) {
                emptyList()
            }
        }

    /**
     * CSV columns (0-indexed):
     * 0=Name, 1=Sex, 2=Event, 3=Equipment, 4=Age, 5=AgeClass, 6=BirthYearClass,
     * 7=Division, 8=BodyweightKg, 9=WeightClassKg,
     * 10-13=Squat1-4Kg, 14=Best3SquatKg,
     * 15-18=Bench1-4Kg, 19=Best3BenchKg,
     * 20-23=Deadlift1-4Kg, 24=Best3DeadliftKg,
     * 25=TotalKg, 26=Place, 27=Dots, 28=Wilks, 29=Glossbrenner, 30=Goodlift,
     * 31=Tested, 32=Country, 33=State, 34=Federation, 35=ParentFederation,
     * 36=Date, 37=MeetCountry, 38=MeetState, 39=MeetTown, 40=MeetName, 41=Sanctioned
     */
    private fun parseCsv(csv: String): List<CompetitionResult> {
        val lines = csv.lines()
        if (lines.size < 2) return emptyList()
        return lines.drop(1)
            .filter { it.isNotBlank() }
            .mapNotNull { line ->
                try {
                    val f = parseCsvLine(line)
                    val date = f.getOrNull(36)?.takeIf { it.isNotEmpty() } ?: return@mapNotNull null
                    val meetName = f.getOrNull(40)?.takeIf { it.isNotEmpty() } ?: return@mapNotNull null
                    CompetitionResult(
                        date = date,
                        meetName = meetName,
                        federation = f.getOrNull(34)?.takeIf { it.isNotEmpty() },
                        equipment = f.getOrNull(3)?.takeIf { it.isNotEmpty() },
                        division = f.getOrNull(7)?.takeIf { it.isNotEmpty() },
                        weightClassKg = f.getOrNull(9)?.takeIf { it.isNotEmpty() },
                        bodyweightKg = f.getOrNull(8)?.toDoubleOrNull(),
                        best3SquatKg = f.getOrNull(14)?.toDoubleOrNull()?.takeIf { it > 0 },
                        best3BenchKg = f.getOrNull(19)?.toDoubleOrNull()?.takeIf { it > 0 },
                        best3DeadliftKg = f.getOrNull(24)?.toDoubleOrNull()?.takeIf { it > 0 },
                        totalKg = f.getOrNull(25)?.toDoubleOrNull()?.takeIf { it > 0 },
                        place = f.getOrNull(26)?.takeIf { it.isNotEmpty() },
                        dots = f.getOrNull(27)?.toDoubleOrNull(),
                        meetCountry = f.getOrNull(37)?.takeIf { it.isNotEmpty() },
                        meetTown = f.getOrNull(39)?.takeIf { it.isNotEmpty() }
                    )
                } catch (e: Exception) {
                    null
                }
            }
    }

    private fun parseCsvLine(line: String): List<String> {
        val result = mutableListOf<String>()
        val field = StringBuilder()
        var inQuotes = false
        for (ch in line) {
            when {
                ch == '"' -> inQuotes = !inQuotes
                ch == ',' && !inQuotes -> {
                    result.add(field.toString())
                    field.clear()
                }
                else -> field.append(ch)
            }
        }
        result.add(field.toString())
        return result
    }

    /**
     * Search a specific gender path, doing up to 3 iterations from startAt.
     * Returns the athletes found and the next start position (null if exhausted).
     */
    private suspend fun searchPath(
        path: String,
        query: String,
        startAt: Int
    ): Pair<List<OplAthlete>, Int?> =
        withContext(Dispatchers.IO) {
            try {
                val accumulated = LinkedHashMap<String, OplAthlete>()
                var start = startAt
                var nextStart: Int? = null

                for (i in 0 until 3) {
                    val searchUrl = "$baseUrl/search/rankings/$path" +
                        "?q=${encode(query)}&start=$start&lang=en&units=kg"
                    val searchResponse = get(searchUrl) ?: break
                    val searchJson = JsonParser.parseString(searchResponse).asJsonObject
                    val nextIndex = searchJson.get("next_index")
                        ?.takeIf { !it.isJsonNull }?.asInt ?: break

                    val rowsUrl = "$baseUrl/rankings/$path" +
                        "?start=$nextIndex&end=${nextIndex + 24}&lang=en&units=kg"
                    val rowsResponse = get(rowsUrl) ?: break
                    val rowsJson = JsonParser.parseString(rowsResponse).asJsonObject
                    val rows = rowsJson.getAsJsonArray("rows") ?: break

                    for (athlete in parseRows(rows)) {
                        accumulated.putIfAbsent(athlete.slug, athlete)
                    }
                    start = nextIndex + 1
                    nextStart = start
                }

                Pair(accumulated.values.toList(), nextStart)
            } catch (e: Exception) {
                Pair(emptyList(), null)
            }
        }

    /**
     * Parse row arrays.
     * Row format: [idx, rank, name, slug, instagram, unknown, country, state,
     *              federation, date, country2, state2, meetPath, gender, equipment,
     *              ageClass, division, bodyweight, weightClass, squat, bench,
     *              deadlift, total, score]
     */
    private fun parseRows(rows: JsonArray): List<OplAthlete> {
        val result = mutableListOf<OplAthlete>()
        for (rowElement in rows) {
            try {
                val row = rowElement.asJsonArray
                val name        = row.getStringOrNull(2) ?: continue
                val slug        = row.getStringOrNull(3) ?: continue
                val country     = row.getStringOrNull(6)
                val federation  = row.getStringOrNull(8)
                val date        = row.getStringOrNull(9)
                val gender      = row.getStringOrNull(13)
                val equipment   = row.getStringOrNull(14)
                val weightClass = row.getStringOrNull(18)
                val squat       = row.getDoubleOrNull(19)
                val bench       = row.getDoubleOrNull(20)
                val deadlift    = row.getDoubleOrNull(21)
                val total       = row.getDoubleOrNull(22)

                result.add(
                    OplAthlete(
                        name = name,
                        slug = slug,
                        country = country,
                        federation = federation,
                        bestSquat = squat,
                        bestBench = bench,
                        bestDeadlift = deadlift,
                        bestTotal = total,
                        weightClass = weightClass,
                        equipment = equipment,
                        lastCompDate = date,
                        gender = gender
                    )
                )
            } catch (e: Exception) {
                // Skip malformed rows
            }
        }
        return result
    }

    private fun JsonArray.getStringOrNull(index: Int): String? {
        val element = this[index] ?: return null
        return if (element.isJsonNull) null else element.asString.takeIf { it.isNotEmpty() }
    }

    private fun JsonArray.getDoubleOrNull(index: Int): Double? {
        val element = this[index] ?: return null
        return if (element.isJsonNull) null
        else try { element.asDouble.takeIf { it != 0.0 } } catch (e: Exception) { null }
    }

    private fun get(url: String): String? {
        val request = Request.Builder().url(url).build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return null
            return response.body?.string()
        }
    }

    private fun encode(query: String): String =
        java.net.URLEncoder.encode(query, "UTF-8")
}

File: ios/MyLiftSquad/Sources/Data/API/OplApiService.swift
import Foundation

struct SearchPage {
    let athletes: [OplAthlete]
    let nextMenStart: Int?
    let nextWomenStart: Int?
    var hasMore: Bool { nextMenStart != nil || nextWomenStart != nil }
}

actor OplApiService {
    static let shared = OplApiService()
    private let baseURL = "https://www.openpowerlifting.org/api"

    private init() {}

    // MARK: - Search

    /// Search for athletes. Pass nextMenStart/nextWomenStart from a previous SearchPage
    /// to load more results. A nil start means that gender path is exhausted and will be skipped.
    func searchAthletes(query: String, menStart: Int? = 0, womenStart: Int? = 0) async throws -> SearchPage {
        let trimmed = query.trimmingCharacters(in: .whitespaces)
        guard trimmed.count >= 2 else { return SearchPage(athletes: [], nextMenStart: nil, nextWomenStart: nil) }

        let encoded = trimmed.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? trimmed

        async let menResult: ([OplAthlete], Int?) = menStart != nil
            ? searchGender("men", query: encoded, startAt: menStart!)
            : ([], nil)
        async let womenResult: ([OplAthlete], Int?) = womenStart != nil
            ? searchGender("women", query: encoded, startAt: womenStart!)
            : ([], nil)

        let (men, nextMenStart) = try await menResult
        let (women, nextWomenStart) = try await womenResult

        var seen = Set<String>()
        var combined: [OplAthlete] = []
        for athlete in (men + women) {
            if !seen.contains(athlete.slug) {
                seen.insert(athlete.slug)
                combined.append(athlete)
            }
        }
        return SearchPage(athletes: combined, nextMenStart: nextMenStart, nextWomenStart: nextWomenStart)
    }

    /// Search a specific gender path, doing up to 3 iterations from startAt.
    /// Returns the athletes found and the next start position (nil if exhausted).
    private func searchGender(_ gender: String, query: String, startAt: Int) async throws -> ([OplAthlete], Int?) {
        let queryLower = query.removingPercentEncoding?.lowercased() ?? query.lowercased()
        var accumulated: [String: OplAthlete] = [:]
        var start = startAt
        var nextStart: Int? = nil

        for _ in 0..<3 {
            let searchURL = "\(baseURL)/search/rankings/\(gender)?q=\(query)&start=\(start)&lang=en&units=kg"
            guard let url = URL(string: searchURL) else { break }

            let (searchData, _) = try await URLSession.shared.data(from: url)
            guard let json = try? JSONSerialization.jsonObject(with: searchData) as? [String: Any],
                  let nextIndex = json["next_index"] as? Int else { break }

            let rankURL = "\(baseURL)/rankings/\(gender)?start=\(nextIndex)&end=\(nextIndex + 24)&lang=en&units=kg"
            guard let rUrl = URL(string: rankURL) else { break }

            let (rankData, _) = try await URLSession.shared.data(from: rUrl)
            guard let rankJson = try? JSONSerialization.jsonObject(with: rankData) as? [String: Any],
                  let rows = rankJson["rows"] as? [[Any]] else { break }

            for row in rows {
                guard row.count >= 24,
                      let name = row[2] as? String,
                      let slug = row[3] as? String,
                      name.lowercased().contains(queryLower),
                      accumulated[slug] == nil else { continue }

                accumulated[slug] = OplAthlete(
                    name: name,
                    slug: slug,
                    country: row[6] as? String ?? "",
                    federation: row[8] as? String ?? "",
                    bestSquatKg: parseDouble(row[19]),
                    bestBenchKg: parseDouble(row[20]),
                    bestDeadliftKg: parseDouble(row[21]),
                    bestTotalKg: parseDouble(row[22]),
                    weightClass: row[18] as? String ?? "",
                    equipment: row[14] as? String ?? "",
                    lastCompDate: row[9] as? String ?? "",
                    gender: gender == "men" ? "M" : "F"
                )
            }

            start = nextIndex + 1
            nextStart = start
        }

        return (Array(accumulated.values), nextStart)
    }

    // MARK: - Competition History

    func fetchHistory(slug: String) async throws -> ([CompetitionResult], OplAthlete?) {
        let csvURL = "\(baseURL)/liftercsv/\(slug)"
        guard let url = URL(string: csvURL) else { return ([], nil) }

        let (data, _) = try await URLSession.shared.data(from: url)
        guard let csvString = String(data: data, encoding: .utf8) else { return ([], nil) }

        let results = parseCSV(csvString, slug: slug)
        let prAthlete = buildAthletePRs(from: results, slug: slug)
        return (results, prAthlete)
    }

    // MARK: - CSV Parsing
    //
    // Column indices (0-based), matching Android OplApiService:
    //  0=Name  1=Sex  2=Event  3=Equipment  4=Age  5=AgeClass  6=BirthYearClass
    //  7=Division  8=BodyweightKg  9=WeightClassKg
    //  10=Squat1  11=Squat2  12=Squat3  13=Squat4  14=Best3SquatKg
    //  15=Bench1  16=Bench2  17=Bench3  18=Bench4  19=Best3BenchKg
    //  20=DL1     21=DL2     22=DL3     23=DL4     24=Best3DeadliftKg
    //  25=TotalKg  26=Place  27=Dots  28=Wilks  29=Glossbrenner  30=Goodlift
    //  31=Tested  32=Country  33=State  34=Federation  35=ParentFederation
    //  36=Date  37=MeetCountry  38=MeetState  39=MeetTown  40=MeetName  41=Sanctioned

    private func parseCSV(_ csv: String, slug: String) -> [CompetitionResult] {
        let lines = csv.components(separatedBy: "\n")
        guard lines.count >= 2 else { return [] }

        let invalidPlaces: Set<String> = ["DQ", "DD", "DNS", "NS", "G"]

        return lines.dropFirst().compactMap { line in
            guard !line.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }
            let f = parseCSVLine(line)

            let date = f.count > 36 ? f[36] : ""
            let meetName = f.count > 40 ? f[40] : ""
            guard !date.isEmpty, !meetName.isEmpty else { return nil }

            let place = f.count > 26 ? f[26] : ""
            guard !invalidPlaces.contains(place) else { return nil }

            return CompetitionResult(
                date: date,
                meetName: meetName,
                federation: f.count > 34 ? f[34] : "",
                equipment: f.count > 3  ? f[3]  : "",
                division:  f.count > 7  ? f[7]  : "",
                weightClassKg: f.count > 9  ? f[9]  : "",
                bodyweightKg:  f.count > 8  ? Double(f[8])  ?? 0 : 0,
                best3SquatKg:  f.count > 14 ? Double(f[14]) ?? 0 : 0,
                best3BenchKg:  f.count > 19 ? Double(f[19]) ?? 0 : 0,
                best3DeadliftKg: f.count > 24 ? Double(f[24]) ?? 0 : 0,
                totalKg:       f.count > 25 ? Double(f[25]) ?? 0 : 0,
                place: place,
                dots:          f.count > 27 ? Double(f[27]) ?? 0 : 0,
                meetCountry:   f.count > 37 ? f[37] : "",
                meetTown:      f.count > 39 ? f[39] : ""
            )
        }.sorted { $0.date > $1.date }
    }

    private func parseCSVLine(_ line: String) -> [String] {
        var fields: [String] = []
        var current = ""
        var inQuotes = false

        for char in line {
            if char == "\"" {
                inQuotes.toggle()
            } else if char == "," && !inQuotes {
                fields.append(current)
                current = ""
            } else {
                current.append(char)
            }
        }
        fields.append(current)
        return fields
    }

    private func buildAthletePRs(from results: [CompetitionResult], slug: String) -> OplAthlete? {
        guard let first = results.first else { return nil }

        var bestSquat: Double = 0
        var bestBench: Double = 0
        var bestDeadlift: Double = 0
        var bestTotal: Double = 0

        for result in results {
            if result.best3SquatKg > bestSquat { bestSquat = result.best3SquatKg }
            if result.best3BenchKg > bestBench { bestBench = result.best3BenchKg }
            if result.best3DeadliftKg > bestDeadlift { bestDeadlift = result.best3DeadliftKg }
            if result.totalKg > bestTotal { bestTotal = result.totalKg }
        }

        return OplAthlete(
            name: "",
            slug: slug,
            country: "",
            federation: first.federation,
            bestSquatKg: bestSquat,
            bestBenchKg: bestBench,
            bestDeadliftKg: bestDeadlift,
            bestTotalKg: bestTotal,
            weightClass: first.weightClassKg,
            equipment: first.equipment,
            lastCompDate: first.date,
            gender: ""
        )
    }

    // MARK: - Helpers

    private func parseDouble(_ value: Any) -> Double {
        if let d = value as? Double { return d }
        if let s = value as? String { return Double(s) ?? 0 }
        if let i = value as? Int { return Double(i) }
        return 0
    }
}
